asynciterator.ts

```diff
@@ -1981,42 +1984,170 @@ class HistoryReader<T> {
   }
 }
 
+/**
+ * An iterator that takes a variety of iterable objects as a source.
+ */
+export class WrappingIterator<T> extends AsyncIterator<T> {
+  protected _source: InternalSource<T> | null = null;
+
+  constructor(source?: MaybePromise<IterableSource<T>>) {
+    super();
+    if (source) {
+      if (!isPromise(source))
+        this.source = source as any;
+      else
+        source.then(s => this.source = s).catch(e => this.emit('error', e));
+    }
+  }
+
+  set source(value: IterableSource<T>) {
+    let source: InternalSource<T> = value as any;
+    // Do not change sources if the iterator is already done
+    if (this.done)
+      return;
+
+    // Process an iterable source
+    if (isIterable(source))
+      source = source[Symbol.iterator]() as any;
+    // Process an iterator source
+    if (isIterator<T>(source)) {
+      let iterator: Iterator<T> | null = source;
+      source = new EventEmitter() as any;
+      source.read = (): T | null => {
+        if (iterator !== null) {
+          // Skip any null values inside of the iterator
+          let next: IteratorResult<T>;
+          while (!(next = iterator.next()).done) {
+            if (next.value !== null)
+              return next.value;
+          }
+          // No remaining values, so stop iterating
+          this.close();
+          iterator = null;
+        }
+        return null;
+      };
+    }
+    // Process any other readable source
+    else {
+      source = ensureSourceAvailable(source);
+    }
+
+    // Set up event handling
+    source._destination = this;
+    source.on('end', destinationClose);
+    source.on('error', destinationEmitError);
+    source.on('readable', destinationSetReadable);
+
+    // Enable reading from source
+    this._source = source;
+    this.readable = true;
```

Comment:
I would prefer `this.readable = source.readable` for the sake of performance; that way we don't trigger reading flows downstream if the source does not have any elements available.

Reply:
We can't guarantee that property to be there, though... Perhaps `this.readable = typeof source.readable === 'boolean' ? source.readable : true`

Reply:
Perfect, I'll combine both of your answers.

Reply:
Are you sure. The lines above rely on it having a readable event?

Reply:
@jeswr  when it comes to the native `stream` module, the `readable` property came with Node.js 11.x if I recall correctly. Up until then we only had the `readable` event. Consequently, there's plenty of old versions of the userland `readable-stream` package out there without it.